crates/evm/src/executor/inspector/cheatcodes/mod.rs

```diff
@@ -538,6 +558,165 @@ impl<DB: DatabaseExt> Inspector<DB> for Cheatcodes {
         InstructionResult::Continue
     }
 
+    #[inline]
+    fn step_end(
+        &mut self,
+        interpreter: &mut Interpreter,
+        data: &mut EVMData<'_, DB>,
+        eval: InstructionResult,
+    ) -> InstructionResult {
+        // for cool cheatcode gas costs,
+        // see https://eips.ethereum.org/EIPS/eip-2200, https://eips.ethereum.org/EIPS/eip-2929
+
+        // if previous step added gas, add it once
+        // note that all the opcodes will already have a cost (usually 100)
+        // so adding until it hits the gas expected for a cold key/address
+        if self.cool_additional_gas_next_op > 0 {
+            interpreter.gas.record_cost(self.cool_additional_gas_next_op);
```

Comment:
main thing that happens is setting the Gas at each op code. since the cost isn't known until the next op code it records it at the next one, and resets. All the other state is just there to track when `vm.cool` is called and each time an opcode is stepped through (when to "consume" the cool state and make it warm) for both the the address and the address' slots